{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '../')\n",
    "from variant.varianthdf5 import VariantHDF5\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Resetting ...\n",
      "Getting data start position ...\n",
      "Making variant HDF5 ...\n",
      "\t@ 1 ...\n",
      "\t\tMaking chromosome 1 variant table ...\n",
      "\t\tMaking chromosome 10 variant table ...\n",
      "\t\tMaking chromosome 11 variant table ...\n",
      "\t\tMaking chromosome 12 variant table ...\n",
      "\t@ 1000001 ...\n",
      "\t\tMaking chromosome 13 variant table ...\n",
      "\t\tMaking chromosome 14 variant table ...\n",
      "\t\tMaking chromosome 15 variant table ...\n",
      "\t\tMaking chromosome 16 variant table ...\n",
      "\t\tMaking chromosome 17 variant table ...\n",
      "\t\tMaking chromosome 18 variant table ...\n",
      "\t@ 2000001 ...\n",
      "\t\tMaking chromosome 19 variant table ...\n",
      "\t\tMaking chromosome 2 variant table ...\n",
      "\t\tMaking chromosome 20 variant table ...\n",
      "\t\tMaking chromosome 21 variant table ...\n",
      "\t\tMaking chromosome 22 variant table ...\n",
      "\t\tMaking chromosome 3 variant table ...\n",
      "\t@ 3000001 ...\n",
      "\t\tMaking chromosome 4 variant table ...\n",
      "\t\tMaking chromosome 5 variant table ...\n",
      "\t\tMaking chromosome 6 variant table ...\n",
      "\t@ 4000001 ...\n",
      "\t\tMaking chromosome 7 variant table ...\n",
      "\t\tMaking chromosome 8 variant table ...\n",
      "\t\tMaking chromosome 9 variant table ...\n",
      "\t\tMaking chromosome MT variant table ...\n",
      "\t\tMaking chromosome X variant table ...\n",
      "\t@ 5000001 ...\n",
      "\t\tMaking chromosome Y variant table ...\n",
      "Flushing tables and making column indices ...\n",
      "\tchromosome 1 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 10 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 11 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 12 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 13 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 14 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 15 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 16 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 17 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 18 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 19 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n",
      "\t\t Making QUAL index ...\n",
      "\t\t Making start index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making variant_type index ...\n",
      "\t\t Making effect index ...\n",
      "\t\t Making impact index ...\n",
      "\t\t Making gene_name index ...\n",
      "\t\t Making GT index ...\n",
      "\tchromosome 2 variant table ...\n",
      "\t\t Making CHROM index ...\n",
      "\t\t Making ID index ...\n",
      "\t\t Making REF index ...\n",
      "\t\t Making ALT index ...\n"
     ]
    }
   ],
   "source": [
    "VARIANT_FILE_PATH = '/home/k/Jumis/guardiome/genome-data/customers/738/variant/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz'\n",
    "variant_hdf5 = VariantHDF5(VARIANT_FILE_PATH, reset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'variant_hdf5' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-e261b35d3b21>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mvariant_hdf5\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_variant_by_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'rs662799'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'variant_hdf5' is not defined"
     ]
    }
   ],
   "source": [
    "variant_hdf5.get_variant_by_id('rs662799')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
