{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (varianthdf5.py, line 132)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"../variant/varianthdf5.py\"\u001b[0;36m, line \u001b[0;32m132\u001b[0m\n\u001b[0;31m    pathogenicity =\u001b[0m\n\u001b[0m                   ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '../')\n",
    "from variant.varianthdf5 import VariantHDF5\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Resetting ...\n",
      "Getting data start position ...\n",
      "Counting variants in chromosomes ...\n",
      "\t@ 1 ...\n",
      "\t@ 10 ...\n",
      "\t@ 11 ...\n",
      "\t@ 12 ...\n",
      "\t@ 13 ...\n",
      "\t@ 14 ...\n",
      "\t@ 15 ...\n",
      "\t@ 16 ...\n",
      "\t@ 17 ...\n",
      "\t@ 18 ...\n",
      "\t@ 19 ...\n",
      "\t@ 2 ...\n",
      "\t@ 20 ...\n",
      "\t@ 21 ...\n",
      "\t@ 22 ...\n",
      "\t@ 3 ...\n",
      "\t@ 4 ...\n",
      "\t@ 5 ...\n",
      "\t@ 6 ...\n",
      "\t@ 7 ...\n",
      "\t@ 8 ...\n",
      "\t@ 9 ...\n",
      "\t@ MT ...\n",
      "\t@ X ...\n",
      "\t@ Y ...\n",
      "defaultdict(<function VariantHDF5._make_variant_hdf5.<locals>.<lambda> at 0x7f0b828311e0>,\n",
      "            {'1': 1,\n",
      "             '10': 1,\n",
      "             '11': 1,\n",
      "             '12': 1,\n",
      "             '13': 1,\n",
      "             '14': 1,\n",
      "             '15': 1,\n",
      "             '16': 1,\n",
      "             '17': 1,\n",
      "             '18': 1,\n",
      "             '19': 1,\n",
      "             '2': 1,\n",
      "             '20': 1,\n",
      "             '21': 1,\n",
      "             '22': 1,\n",
      "             '3': 1,\n",
      "             '4': 1,\n",
      "             '5': 1,\n",
      "             '6': 1,\n",
      "             '7': 1,\n",
      "             '8': 1,\n",
      "             '9': 1,\n",
      "             'MT': 1,\n",
      "             'X': 1,\n",
      "             'Y': 1})\n",
      "Making variant HDF5 ...\n",
      "\t@ 1 ...\n",
      "\t\tMaking chromosome 1 variant table ...\n",
      "\t\tMaking chromosome 10 variant table ...\n",
      "\t\tMaking chromosome 11 variant table ...\n",
      "\t\tMaking chromosome 12 variant table ...\n",
      "\t@ 1000001 ...\n",
      "\t\tMaking chromosome 13 variant table ...\n",
      "\t\tMaking chromosome 14 variant table ...\n",
      "\t\tMaking chromosome 15 variant table ...\n",
      "\t\tMaking chromosome 16 variant table ...\n",
      "\t\tMaking chromosome 17 variant table ...\n",
      "\t\tMaking chromosome 18 variant table ...\n",
      "\t@ 2000001 ...\n",
      "\t\tMaking chromosome 19 variant table ...\n",
      "\t\tMaking chromosome 2 variant table ...\n",
      "\t\tMaking chromosome 20 variant table ...\n",
      "\t\tMaking chromosome 21 variant table ...\n",
      "\t\tMaking chromosome 22 variant table ...\n",
      "\t\tMaking chromosome 3 variant table ...\n",
      "\t@ 3000001 ...\n",
      "\t\tMaking chromosome 4 variant table ...\n",
      "\t\tMaking chromosome 5 variant table ...\n",
      "\t\tMaking chromosome 6 variant table ...\n",
      "\t@ 4000001 ...\n",
      "\t\tMaking chromosome 7 variant table ...\n",
      "\t\tMaking chromosome 8 variant table ...\n",
      "\t\tMaking chromosome 9 variant table ...\n",
      "\t\tMaking chromosome MT variant table ...\n",
      "\t\tMaking chromosome X variant table ...\n",
      "\t@ 5000001 ...\n",
      "\t\tMaking chromosome Y variant table ...\n",
      "Flushing tables and making column indices ...\n",
      "\tchromosome 1 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 10 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 11 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 12 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 13 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 14 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 15 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 16 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 17 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 18 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 19 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 2 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 20 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 21 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 22 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 3 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 4 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 5 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 6 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 7 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 8 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 9 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome MT variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome X variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome Y variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "/home/k/Jumis/guardiome/genome-data/customers/738/variant/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz.hdf5 (File) ''\n",
      "Last modif.: 'Fri May 26 01:07:21 2017'\n",
      "Object Tree: \n",
      "/ (RootGroup) ''\n",
      "/chromosome_10_variants (Table(249020,), shuffle, blosc(1)) ''\n",
      "/chromosome_11_variants (Table(240544,), shuffle, blosc(1)) ''\n",
      "/chromosome_12_variants (Table(231845,), shuffle, blosc(1)) ''\n",
      "/chromosome_13_variants (Table(179803,), shuffle, blosc(1)) ''\n",
      "/chromosome_14_variants (Table(152565,), shuffle, blosc(1)) ''\n",
      "/chromosome_15_variants (Table(146512,), shuffle, blosc(1)) ''\n",
      "/chromosome_16_variants (Table(159467,), shuffle, blosc(1)) ''\n",
      "/chromosome_17_variants (Table(150092,), shuffle, blosc(1)) ''\n",
      "/chromosome_18_variants (Table(138271,), shuffle, blosc(1)) ''\n",
      "/chromosome_19_variants (Table(109788,), shuffle, blosc(1)) ''\n",
      "/chromosome_1_variants (Table(372124,), shuffle, blosc(1)) ''\n",
      "/chromosome_20_variants (Table(144307,), shuffle, blosc(1)) ''\n",
      "/chromosome_21_variants (Table(104367,), shuffle, blosc(1)) ''\n",
      "/chromosome_22_variants (Table(107316,), shuffle, blosc(1)) ''\n",
      "/chromosome_2_variants (Table(401037,), shuffle, blosc(1)) ''\n",
      "/chromosome_3_variants (Table(336097,), shuffle, blosc(1)) ''\n",
      "/chromosome_4_variants (Table(355664,), shuffle, blosc(1)) ''\n",
      "/chromosome_5_variants (Table(285518,), shuffle, blosc(1)) ''\n",
      "/chromosome_6_variants (Table(296731,), shuffle, blosc(1)) ''\n",
      "/chromosome_7_variants (Table(289264,), shuffle, blosc(1)) ''\n",
      "/chromosome_8_variants (Table(243301,), shuffle, blosc(1)) ''\n",
      "/chromosome_9_variants (Table(236641,), shuffle, blosc(1)) ''\n",
      "/chromosome_MT_variants (Table(30,), shuffle, blosc(1)) ''\n",
      "/chromosome_X_variants (Table(106614,), shuffle, blosc(1)) ''\n",
      "/chromosome_Y_variants (Table(23120,), shuffle, blosc(1)) ''\n",
      "\n",
      "Writing RSID-to-chromosome dict ...\n"
     ]
    }
   ],
   "source": [
    "VARIANT_FILE_PATH = '/home/k/Jumis/guardiome/genome-data/customers/738/variant/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz'\n",
    "variant_hdf5 = VariantHDF5(VARIANT_FILE_PATH, reset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(b'A', b'11', 116792991, b'', b'rs662799', 1181.77001953125, b'G', 116792991)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_hdf5.get_variant_by_id('rs662799')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
