{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pickle import *\n",
    "from pprint import pprint\n",
    "from gzip import *\n",
    "from tables import *\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_variant_start_and_end_positions(pos, ref, alt):\n",
    "    \"\"\"\n",
    "    Get variant (@pos: ref ==> alt) start and end positions.\n",
    "    :param pos: int;\n",
    "    :param ref: str;\n",
    "    :param alt: str;\n",
    "    :return: (str, str);\n",
    "    \"\"\"\n",
    "\n",
    "    # TODO: Check accuracy\n",
    "\n",
    "    if len(ref) == len(alt):\n",
    "        s, e = pos, pos + len(alt) - 1\n",
    "\n",
    "    elif len(ref) < len(alt):\n",
    "        s, e = pos, pos + 1\n",
    "\n",
    "    else:  # len(alt) < len(ref)\n",
    "        s, e = pos + 1, pos + len(ref) - len(alt)\n",
    "\n",
    "    return s, e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class VariantHDF5:\n",
    "    \"\"\"\n",
    "    Data structure storing variants in .HDF5.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, variant_file_path, reset=False):\n",
    "        \"\"\"\n",
    "        Construct VariantHDF5.\n",
    "        :param reset: bool; whether to re-make data instead of reading from files\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "\n",
    "        # File paths\n",
    "        self.variant_file_path = variant_file_path\n",
    "        self.variant_hdf5_file_path = '{}.hdf5'.format(self.variant_file_path)\n",
    "        self.variant_rsid_to_chrom_dict_file_path = '{}.rsid_to_chrom_dict.pickle.gz'.format(\n",
    "            self.variant_file_path)\n",
    "\n",
    "        # Data\n",
    "        self.variant_hdf5 = None\n",
    "        self.rsid_to_chrom_dict = {}\n",
    "\n",
    "        self._load_data(reset=reset)\n",
    "\n",
    "    def __del__(self):\n",
    "        \"\"\"\n",
    "        Destruct VariantHDF5.\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "\n",
    "        if self.variant_hdf5:\n",
    "            self.variant_hdf5.close()\n",
    "            print('Closed variant HDF5.')\n",
    "\n",
    "    def _load_data(self, reset=False):\n",
    "        \"\"\"\n",
    "        Initialize self.variant_hdf5 & self.rsid_to_chrom_dict.\n",
    "        :param reset: bool; whether to re-make data instead of reading from files\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "        \n",
    "        if not reset:\n",
    "\n",
    "            try:\n",
    "                print('Reading variant HDF5 ...')\n",
    "                self.variant_hdf5 = open_file(\n",
    "                    self.variant_hdf5_file_path, mode='r')\n",
    "\n",
    "                print('Reading RSID-to-chromosome dict ...')\n",
    "                self._read_rsid_to_chrom_dict()\n",
    "\n",
    "            except (FileNotFoundError, HDF5ExtError) as e:\n",
    "                print('\\tFailed ({}).'.format(e))\n",
    "                reset = True\n",
    "\n",
    "        if reset:\n",
    "            print('Resetting ...')\n",
    "            \n",
    "            if self.variant_hdf5:\n",
    "                self.variant_hdf5.close()\n",
    "                \n",
    "            self._make_variant_hdf5()\n",
    "            \n",
    "            self.variant_hdf5 = open_file(\n",
    "                self.variant_hdf5_file_path, mode='r')\n",
    "\n",
    "    def _make_variant_hdf5(self):\n",
    "        \"\"\"\n",
    "        Make .HDF5 storing variants.\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "\n",
    "        with open(self.variant_file_path, 'rt') as f:\n",
    "\n",
    "            print('Getting data start position ...')\n",
    "            data_start_position = None\n",
    "            line = f.readline()\n",
    "            while line.startswith('#'):\n",
    "                data_start_position = f.tell()\n",
    "                line = f.readline()\n",
    "\n",
    "            print('Counting variants in chromosomes ...')\n",
    "            chrom_n_rows = defaultdict(lambda: 0)\n",
    "            chrom = None\n",
    "            while line:\n",
    "                a_chrom = line.split('\\t')[0]\n",
    "                if chrom != a_chrom:\n",
    "                    print('\\t@ {} ...'.format(a_chrom))\n",
    "                    chrom = a_chrom\n",
    "                    chrom_n_rows[a_chrom] += 1\n",
    "                line = f.readline()\n",
    "            pprint(chrom_n_rows)\n",
    "\n",
    "            print('Making variant HDF5 ...')\n",
    "            with open_file(\n",
    "                    self.variant_hdf5_file_path,\n",
    "                    mode='w',\n",
    "                    filters=Filters(\n",
    "                        complevel=1, complib='blosc')) as variant_hdf5:\n",
    "\n",
    "                chrom_table_to_row_dict = {}\n",
    "\n",
    "                f.seek(data_start_position)\n",
    "                for i, line in enumerate(f):\n",
    "                    if i % 1000000 == 0:\n",
    "                        print('\\t@ {} ...'.format(i + 1))\n",
    "\n",
    "                    chrom, pos, id_, ref, alt, qual, filter_, info, format_, sample = line.split(\n",
    "                        '\\t')[:10]\n",
    "                    start, end = get_variant_start_and_end_positions(\n",
    "                        int(pos), ref, alt)\n",
    "\n",
    "                    if chrom not in chrom_table_to_row_dict:\n",
    "                        chrom_table = variant_hdf5.create_table(\n",
    "                            '/',\n",
    "                            'chromosome_{}_variants'.format(chrom),\n",
    "                            description=self._VariantDescription,\n",
    "                            #                             expectedrows=chrom_n_rows[chrom],\n",
    "                        )\n",
    "                        print('\\t\\tMaking chromosome {} variant table ...'.\n",
    "                              format(chrom))\n",
    "                        chrom_table_to_row_dict[chrom] = chrom_table.row\n",
    "\n",
    "                    cursor = chrom_table_to_row_dict[chrom]\n",
    "                    cursor['chrom'] = chrom\n",
    "                    cursor['start'] = start\n",
    "                    cursor['end'] = end\n",
    "                    cursor['id_'] = id_\n",
    "                    cursor['ref'] = ref\n",
    "                    cursor['alt'] = alt\n",
    "                    cursor['qual'] = qual\n",
    "\n",
    "                    if id_ != '.':\n",
    "                        self.rsid_to_chrom_dict[id_] = chrom\n",
    "\n",
    "                    cursor.append()\n",
    "\n",
    "                print('Flushing tables and making column indices ...')\n",
    "                for chrom in chrom_table_to_row_dict:\n",
    "                    print('\\tchromosome {} variant table ...'.format(chrom))\n",
    "                    chrom_table = variant_hdf5.get_node(\n",
    "                        '/', 'chromosome_{}_variants'.format(chrom))\n",
    "                    chrom_table.flush()\n",
    "\n",
    "                    for col in sorted(\n",
    "                        ['chrom', 'start', 'end', 'id_', 'ref', 'alt']):\n",
    "                        print('\\t\\t Making {} index ...'.format(col))\n",
    "                        chrom_table.cols._f_col(col).create_csindex()\n",
    "\n",
    "                self.variant_hdf5 = variant_hdf5\n",
    "                print(self.variant_hdf5)\n",
    "\n",
    "                print('Writing RSID-to-chromosome dict ...')\n",
    "                self._write_rsid_to_chrom_dict()\n",
    "\n",
    "    class _VariantDescription(IsDescription):\n",
    "        \"\"\"\n",
    "        Describe VariantHDF5 table columns.\n",
    "        \"\"\"\n",
    "\n",
    "        # TODO: Match with VCF specification\n",
    "        chrom = StringCol(16)\n",
    "        start = Int32Col()\n",
    "        end = Int32Col()\n",
    "        id_ = StringCol(16)\n",
    "        ref = StringCol(256)\n",
    "        alt = StringCol(256)\n",
    "        qual = Float32Col()\n",
    "        gt = StringCol(16)\n",
    "\n",
    "    def _write_rsid_to_chrom_dict(self):\n",
    "        \"\"\"\n",
    "        Write RSID-to-chromosome dict to file.\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "\n",
    "        with open(self.variant_rsid_to_chrom_dict_file_path, 'wb') as f:\n",
    "            dump(self.rsid_to_chrom_dict, f)\n",
    "\n",
    "    def _read_rsid_to_chrom_dict(self):\n",
    "        \"\"\"\n",
    "        Read RSID-to-chromosome dict from file.\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "\n",
    "        with open(self.variant_rsid_to_chrom_dict_file_path, 'rb') as f:\n",
    "            self.rsid_to_chrom_dict = load(f)\n",
    "\n",
    "    def get_variant_by_id(self, id_):\n",
    "        \"\"\"\n",
    "        Search for id_ in variants.\n",
    "        :param id_: str; RSID\n",
    "        :return: dict; {}\n",
    "        \"\"\"\n",
    "\n",
    "        chrom = self.rsid_to_chrom_dict.get(id_)\n",
    "\n",
    "        if chrom:  # Variant found\n",
    "            chrom_table = self.variant_hdf5.get_node(\n",
    "                '/', 'chromosome_{}_variants'.format(chrom))\n",
    "            return [\n",
    "                row.fetch_all_fields()\n",
    "                for row in chrom_table.where(\"id_ == b'{}'\".format(id_))\n",
    "            ]\n",
    "\n",
    "        else:  # Variant not found\n",
    "            return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Resetting ...\n",
      "Getting data start position ...\n",
      "Counting variants in chromosomes ...\n",
      "\t@ 1 ...\n",
      "\t@ 10 ...\n",
      "\t@ 11 ...\n",
      "\t@ 12 ...\n",
      "\t@ 13 ...\n",
      "\t@ 14 ...\n",
      "\t@ 15 ...\n",
      "\t@ 16 ...\n",
      "\t@ 17 ...\n",
      "\t@ 18 ...\n",
      "\t@ 19 ...\n",
      "\t@ 2 ...\n",
      "\t@ 20 ...\n",
      "\t@ 21 ...\n",
      "\t@ 22 ...\n",
      "\t@ 3 ...\n",
      "\t@ 4 ...\n",
      "\t@ 5 ...\n",
      "\t@ 6 ...\n",
      "\t@ 7 ...\n",
      "\t@ 8 ...\n",
      "\t@ 9 ...\n",
      "\t@ MT ...\n",
      "\t@ X ...\n",
      "\t@ Y ...\n",
      "defaultdict(<function VariantHDF5._make_variant_hdf5.<locals>.<lambda> at 0x7fee86f01d90>,\n",
      "            {'1': 1,\n",
      "             '10': 1,\n",
      "             '11': 1,\n",
      "             '12': 1,\n",
      "             '13': 1,\n",
      "             '14': 1,\n",
      "             '15': 1,\n",
      "             '16': 1,\n",
      "             '17': 1,\n",
      "             '18': 1,\n",
      "             '19': 1,\n",
      "             '2': 1,\n",
      "             '20': 1,\n",
      "             '21': 1,\n",
      "             '22': 1,\n",
      "             '3': 1,\n",
      "             '4': 1,\n",
      "             '5': 1,\n",
      "             '6': 1,\n",
      "             '7': 1,\n",
      "             '8': 1,\n",
      "             '9': 1,\n",
      "             'MT': 1,\n",
      "             'X': 1,\n",
      "             'Y': 1})\n",
      "Making variant HDF5 ...\n",
      "\t@ 1 ...\n",
      "\t\tMaking chromosome 1 variant table ...\n",
      "\t\tMaking chromosome 10 variant table ...\n",
      "\t\tMaking chromosome 11 variant table ...\n",
      "\t\tMaking chromosome 12 variant table ...\n",
      "\t@ 1000001 ...\n",
      "\t\tMaking chromosome 13 variant table ...\n",
      "\t\tMaking chromosome 14 variant table ...\n",
      "\t\tMaking chromosome 15 variant table ...\n",
      "\t\tMaking chromosome 16 variant table ...\n",
      "\t\tMaking chromosome 17 variant table ...\n",
      "\t\tMaking chromosome 18 variant table ...\n",
      "\t@ 2000001 ...\n",
      "\t\tMaking chromosome 19 variant table ...\n",
      "\t\tMaking chromosome 2 variant table ...\n",
      "\t\tMaking chromosome 20 variant table ...\n",
      "\t\tMaking chromosome 21 variant table ...\n",
      "\t\tMaking chromosome 22 variant table ...\n",
      "\t\tMaking chromosome 3 variant table ...\n",
      "\t@ 3000001 ...\n",
      "\t\tMaking chromosome 4 variant table ...\n",
      "\t\tMaking chromosome 5 variant table ...\n",
      "\t\tMaking chromosome 6 variant table ...\n",
      "\t@ 4000001 ...\n",
      "\t\tMaking chromosome 7 variant table ...\n",
      "\t\tMaking chromosome 8 variant table ...\n",
      "\t\tMaking chromosome 9 variant table ...\n",
      "\t\tMaking chromosome MT variant table ...\n",
      "\t\tMaking chromosome X variant table ...\n",
      "\t@ 5000001 ...\n",
      "\t\tMaking chromosome Y variant table ...\n",
      "Flushing tables and making column indices ...\n",
      "\tchromosome 1 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 10 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 11 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 12 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 13 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 14 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 15 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 16 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 17 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 18 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 19 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 2 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 20 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 21 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 22 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 3 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 4 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 5 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 6 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 7 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 8 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome 9 variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome MT variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome X variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "\tchromosome Y variant table ...\n",
      "\t\t Making alt index ...\n",
      "\t\t Making chrom index ...\n",
      "\t\t Making end index ...\n",
      "\t\t Making id_ index ...\n",
      "\t\t Making ref index ...\n",
      "\t\t Making start index ...\n",
      "/home/k/Jumis/guardiome/genome-data/customers/738/variant/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz.hdf5 (File) ''\n",
      "Last modif.: 'Thu May 25 11:34:31 2017'\n",
      "Object Tree: \n",
      "/ (RootGroup) ''\n",
      "/chromosome_10_variants (Table(249020,), shuffle, blosc(1)) ''\n",
      "/chromosome_11_variants (Table(240544,), shuffle, blosc(1)) ''\n",
      "/chromosome_12_variants (Table(231845,), shuffle, blosc(1)) ''\n",
      "/chromosome_13_variants (Table(179803,), shuffle, blosc(1)) ''\n",
      "/chromosome_14_variants (Table(152565,), shuffle, blosc(1)) ''\n",
      "/chromosome_15_variants (Table(146512,), shuffle, blosc(1)) ''\n",
      "/chromosome_16_variants (Table(159467,), shuffle, blosc(1)) ''\n",
      "/chromosome_17_variants (Table(150092,), shuffle, blosc(1)) ''\n",
      "/chromosome_18_variants (Table(138271,), shuffle, blosc(1)) ''\n",
      "/chromosome_19_variants (Table(109788,), shuffle, blosc(1)) ''\n",
      "/chromosome_1_variants (Table(372124,), shuffle, blosc(1)) ''\n",
      "/chromosome_20_variants (Table(144307,), shuffle, blosc(1)) ''\n",
      "/chromosome_21_variants (Table(104367,), shuffle, blosc(1)) ''\n",
      "/chromosome_22_variants (Table(107316,), shuffle, blosc(1)) ''\n",
      "/chromosome_2_variants (Table(401037,), shuffle, blosc(1)) ''\n",
      "/chromosome_3_variants (Table(336097,), shuffle, blosc(1)) ''\n",
      "/chromosome_4_variants (Table(355664,), shuffle, blosc(1)) ''\n",
      "/chromosome_5_variants (Table(285518,), shuffle, blosc(1)) ''\n",
      "/chromosome_6_variants (Table(296731,), shuffle, blosc(1)) ''\n",
      "/chromosome_7_variants (Table(289264,), shuffle, blosc(1)) ''\n",
      "/chromosome_8_variants (Table(243301,), shuffle, blosc(1)) ''\n",
      "/chromosome_9_variants (Table(236641,), shuffle, blosc(1)) ''\n",
      "/chromosome_MT_variants (Table(30,), shuffle, blosc(1)) ''\n",
      "/chromosome_X_variants (Table(106614,), shuffle, blosc(1)) ''\n",
      "/chromosome_Y_variants (Table(23120,), shuffle, blosc(1)) ''\n",
      "\n",
      "Writing RSID-to-chromosome dict ...\n"
     ]
    }
   ],
   "source": [
    "VARIANT_FILE_PATH = '/home/k/Jumis/guardiome/genome-data/customers/738/variant/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz'\n",
    "variant_hdf5 = VariantHDF5(VARIANT_FILE_PATH, reset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(b'T', b'1', 182162445, b'rs1573247', 1448.77001953125, b'C', 182162445)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_hdf5.get_variant_by_id('rs1573247')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "REFERENCE_VARIANT_FILE_PATH = '/home/k/Jumis/guardiome/genome-data/grch/38/variants/Homo_sapiens.vcf.gz'\n",
    "REFERENCE_VARIANT_HDF5_FILE_PATH = '{}.hdf5'.format(REFERENCE_VARIANT_FILE_PATH)\n",
    "\n",
    "class ReferenceVariants(IsDescription):\n",
    "    \"\"\"\n",
    "    Column descriptor.\n",
    "    \"\"\"\n",
    "\n",
    "    # TODO: Match with VCF specification\n",
    "    chrom = StringCol(16)\n",
    "    start = Int32Col()\n",
    "    end = Int32Col()\n",
    "    id_ = StringCol(16)\n",
    "    ref = StringCol(256)\n",
    "    alt = StringCol(256)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
