{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tables import *\n",
    "from gzip import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "REFERENCE_VARIANTS_FILE_PATH = './Homo_sapiens.vcf.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class ReferenceVariants(IsDescription):\n",
    "    \"\"\"\n",
    "    Column descriptor.\n",
    "    \"\"\"\n",
    "   \n",
    "    # TODO: Match with VCF specification\n",
    "    chrom = StringCol(16)\n",
    "    start = Int32Col()\n",
    "    end = Int32Col()\n",
    "    id_ = StringCol(16)\n",
    "    ref = StringCol(256)\n",
    "    alt = StringCol(256)\n",
    "\n",
    "    \n",
    "def get_variant_start_and_end_positions(pos, ref, alt):\n",
    "    \"\"\"\n",
    "    Get variant (@pos: ref ==> alt) start and end positions.\n",
    "    :param pos: int;\n",
    "    :param ref: str;\n",
    "    :param alt: str;\n",
    "    :return: (str, str);\n",
    "    \"\"\"\n",
    "    \n",
    "    # TODO: Check accuracy\n",
    "\n",
    "    if len(ref) == len(alt):\n",
    "        s, e = pos, pos + len(alt) - 1\n",
    "\n",
    "    elif len(ref) < len(alt):\n",
    "        s, e = pos, pos + 1\n",
    "\n",
    "    else:  # len(alt) < len(ref)\n",
    "        s, e = pos + 1, pos + len(ref) - len(alt)\n",
    "\n",
    "    return s, e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create HDF5\n",
    "hdf5 = open_file('table.hdf5', mode='w', title='Reference Variants')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create chromosome group\n",
    "group = hdf5.create_group('/', 'chromosome_1', title='Chromosome 1')\n",
    "\n",
    "# Create chromosome table\n",
    "table = hdf5.create_table(group, 'reference_variants', ReferenceVariants, title='Chromosome 1 Reference Variants')\n",
    "print(hdf5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing variant 1000000 ...\n",
      "Processing variant 2000000 ...\n",
      "Processing variant 3000000 ...\n",
      "Processing variant 4000000 ...\n",
      "Processing variant 5000000 ...\n",
      "Processing variant 6000000 ...\n",
      "Processing variant 7000000 ...\n",
      "Processing variant 8000000 ...\n",
      "Processing variant 9000000 ...\n",
      "Processing variant 10000000 ...\n",
      "Processing variant 11000000 ...\n",
      "Processing variant 12000000 ...\n",
      "Processing variant 13000000 ...\n",
      "Processing variant 14000000 ...\n",
      "Processing variant 15000000 ...\n",
      "Processing variant 16000000 ...\n",
      "Processing variant 17000000 ...\n",
      "Processing variant 18000000 ...\n",
      "Processing variant 19000000 ...\n",
      "Processing variant 20000000 ...\n",
      "Processing variant 21000000 ...\n",
      "Processing variant 22000000 ...\n",
      "Processing variant 23000000 ...\n",
      "Processing variant 24000000 ...\n",
      "Processing variant 25000000 ...\n",
      "Processing variant 26000000 ...\n",
      "Processing variant 27000000 ...\n",
      "Processing variant 28000000 ...\n",
      "Processing variant 29000000 ...\n",
      "Processing variant 30000000 ...\n",
      "Processing variant 31000000 ...\n",
      "Processing variant 32000000 ...\n",
      "Processing variant 33000000 ...\n",
      "Processing variant 34000000 ...\n",
      "Processing variant 35000000 ...\n",
      "Processing variant 36000000 ...\n",
      "Processing variant 37000000 ...\n",
      "Processing variant 38000000 ...\n",
      "Processing variant 39000000 ...\n",
      "Processing variant 40000000 ...\n",
      "Processing variant 41000000 ...\n",
      "Processing variant 42000000 ...\n",
      "Processing variant 43000000 ...\n",
      "Processing variant 44000000 ...\n",
      "Processing variant 45000000 ...\n",
      "Processing variant 46000000 ...\n",
      "Processing variant 47000000 ...\n",
      "Processing variant 48000000 ...\n",
      "Processing variant 49000000 ...\n",
      "Processing variant 50000000 ...\n",
      "Processing variant 51000000 ...\n",
      "Processing variant 52000000 ...\n",
      "Processing variant 53000000 ...\n",
      "Processing variant 54000000 ...\n",
      "Processing variant 55000000 ...\n",
      "Processing variant 56000000 ...\n",
      "Processing variant 57000000 ...\n",
      "Processing variant 58000000 ...\n",
      "Processing variant 59000000 ...\n",
      "Processing variant 60000000 ...\n",
      "Processing variant 61000000 ...\n",
      "Processing variant 62000000 ...\n",
      "Processing variant 63000000 ...\n",
      "Processing variant 64000000 ...\n",
      "Processing variant 65000000 ...\n",
      "Processing variant 66000000 ...\n",
      "Processing variant 67000000 ...\n",
      "Processing variant 68000000 ...\n",
      "Processing variant 69000000 ...\n",
      "Processing variant 70000000 ...\n",
      "Processing variant 71000000 ...\n",
      "Processing variant 72000000 ...\n",
      "Processing variant 73000000 ...\n",
      "Processing variant 74000000 ...\n",
      "Processing variant 75000000 ...\n",
      "Processing variant 76000000 ...\n",
      "Processing variant 77000000 ...\n",
      "Processing variant 78000000 ...\n",
      "Processing variant 79000000 ...\n",
      "Processing variant 80000000 ...\n",
      "Processing variant 81000000 ...\n",
      "Processing variant 82000000 ...\n",
      "Processing variant 83000000 ...\n",
      "Processing variant 84000000 ...\n",
      "Processing variant 85000000 ...\n",
      "Processing variant 86000000 ...\n",
      "Processing variant 87000000 ...\n",
      "Processing variant 88000000 ...\n",
      "Processing variant 89000000 ...\n",
      "Processing variant 90000000 ...\n",
      "Processing variant 91000000 ...\n",
      "Processing variant 92000000 ...\n",
      "Processing variant 93000000 ...\n",
      "Processing variant 94000000 ...\n",
      "Processing variant 95000000 ...\n",
      "Processing variant 96000000 ...\n",
      "Processing variant 97000000 ...\n",
      "Processing variant 98000000 ...\n",
      "Processing variant 99000000 ...\n",
      "Processing variant 100000000 ...\n",
      "Processing variant 101000000 ...\n",
      "Processing variant 102000000 ...\n",
      "Processing variant 103000000 ...\n",
      "Processing variant 104000000 ...\n",
      "Processing variant 105000000 ...\n",
      "Processing variant 106000000 ...\n",
      "Processing variant 107000000 ...\n",
      "Processing variant 108000000 ...\n",
      "Processing variant 109000000 ...\n",
      "Processing variant 110000000 ...\n",
      "Processing variant 111000000 ...\n",
      "Processing variant 112000000 ...\n",
      "Processing variant 113000000 ...\n",
      "Processing variant 114000000 ...\n",
      "Processing variant 115000000 ...\n",
      "Processing variant 116000000 ...\n",
      "Processing variant 117000000 ...\n",
      "Processing variant 118000000 ...\n",
      "Processing variant 119000000 ...\n",
      "Processing variant 120000000 ...\n",
      "Processing variant 121000000 ...\n",
      "Processing variant 122000000 ...\n",
      "Processing variant 123000000 ...\n",
      "Processing variant 124000000 ...\n",
      "Processing variant 125000000 ...\n",
      "Processing variant 126000000 ...\n",
      "Processing variant 127000000 ...\n",
      "Processing variant 128000000 ...\n",
      "Processing variant 129000000 ...\n",
      "Processing variant 130000000 ...\n",
      "Processing variant 131000000 ...\n",
      "Processing variant 132000000 ...\n",
      "Processing variant 133000000 ...\n",
      "Processing variant 134000000 ...\n",
      "Processing variant 135000000 ...\n",
      "Processing variant 136000000 ...\n",
      "Processing variant 137000000 ...\n",
      "Processing variant 138000000 ...\n",
      "Processing variant 139000000 ...\n",
      "Processing variant 140000000 ...\n",
      "Processing variant 141000000 ...\n",
      "Processing variant 142000000 ...\n",
      "Processing variant 143000000 ...\n",
      "Processing variant 144000000 ...\n",
      "Processing variant 145000000 ...\n",
      "Processing variant 146000000 ...\n",
      "Processing variant 147000000 ...\n",
      "Processing variant 148000000 ...\n",
      "Processing variant 149000000 ...\n",
      "Processing variant 150000000 ...\n",
      "Processing variant 151000000 ...\n",
      "Processing variant 152000000 ...\n",
      "Processing variant 153000000 ...\n"
     ]
    }
   ],
   "source": [
    "# Populate chromosome table\n",
    "variant = table.row\n",
    "with open(REFERENCE_VARIANT_FILE_PATH) as f:\n",
    "    i = 0\n",
    "    for row in f:\n",
    "        \n",
    "        row = row.decode()\n",
    "        \n",
    "        if row.startswith('#'):\n",
    "            continue\n",
    "            \n",
    "        i += 1\n",
    "        if i % 1000000 == 0:\n",
    "            print('Processing variant {} ...'.format(i))\n",
    "        \n",
    "        chrom, pos, id_, ref, alt, qual, filter_, info = row.split('\\t')[:8] \n",
    "        start, end = get_variant_start_and_end_positions(int(pos), ref, alt)\n",
    "        \n",
    "        variant['chrom'] = chrom\n",
    "        variant['start'] = start\n",
    "        variant['end'] = end\n",
    "        variant['id_'] = id_\n",
    "        variant['ref'] = ref\n",
    "        variant['alt'] = alt\n",
    "        variant.append()\n",
    "\n",
    "table.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "hdf5.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
