{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '../')\n",
    "from variant.varianthdf5 import VariantHDF5\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading variant HDF5 ...\n",
      "Reading ID-to-chromosome dict ...\n",
      "Reading gene-to-chromosome dict ...\n",
      "\tFailed: [Errno 2] No such file or directory: '/home/k/Jumis/guardiome/genome-data/customers/738/variant/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz.gene_to_chrom_dict.pickle.gz'.\n",
      "Resetting ...\n",
      "Closed varinat HDF5.\n",
      "Making variant HDF5 ...\n",
      "Getting data start position ...\n",
      "Counting variants in chromosomes ...\n",
      "\t@ 1 ...\n",
      "\t@ 10 ...\n",
      "\t@ 11 ...\n",
      "\t@ 12 ...\n",
      "\t@ 13 ...\n",
      "\t@ 14 ...\n",
      "\t@ 15 ...\n",
      "\t@ 16 ...\n",
      "\t@ 17 ...\n",
      "\t@ 18 ...\n",
      "\t@ 19 ...\n",
      "\t@ 2 ...\n",
      "\t@ 20 ...\n",
      "\t@ 21 ...\n",
      "\t@ 22 ...\n",
      "\t@ 3 ...\n",
      "\t@ 4 ...\n",
      "\t@ 5 ...\n",
      "\t@ 6 ...\n",
      "\t@ 7 ...\n",
      "\t@ 8 ...\n",
      "\t@ 9 ...\n",
      "\t@ MT ...\n",
      "\t@ X ...\n",
      "\t@ Y ...\n",
      "defaultdict(<function VariantHDF5._make_variant_hdf5.<locals>.<lambda> at 0x7f561255e598>,\n",
      "            {'1': 372124,\n",
      "             '10': 249020,\n",
      "             '11': 240544,\n",
      "             '12': 231845,\n",
      "             '13': 179803,\n",
      "             '14': 152565,\n",
      "             '15': 146512,\n",
      "             '16': 159467,\n",
      "             '17': 150092,\n",
      "             '18': 138271,\n",
      "             '19': 109788,\n",
      "             '2': 401037,\n",
      "             '20': 144307,\n",
      "             '21': 104367,\n",
      "             '22': 107316,\n",
      "             '3': 336097,\n",
      "             '4': 355664,\n",
      "             '5': 285518,\n",
      "             '6': 296731,\n",
      "             '7': 289264,\n",
      "             '8': 243301,\n",
      "             '9': 236641,\n",
      "             'MT': 30,\n",
      "             'X': 106614,\n",
      "             'Y': 23120})\n",
      "Making variant HDF5 ...\n",
      "\t@ 1 ...\n",
      "\t\tMaking chromosome 1 variant table ...\n",
      "\t\tMaking chromosome 10 variant table ...\n",
      "\t\tMaking chromosome 11 variant table ...\n",
      "\t\tMaking chromosome 12 variant table ...\n",
      "\t@ 1000001 ...\n",
      "\t\tMaking chromosome 13 variant table ...\n",
      "\t\tMaking chromosome 14 variant table ...\n",
      "\t\tMaking chromosome 15 variant table ...\n",
      "\t\tMaking chromosome 16 variant table ...\n",
      "\t\tMaking chromosome 17 variant table ...\n",
      "\t\tMaking chromosome 18 variant table ...\n",
      "\t@ 2000001 ...\n",
      "\t\tMaking chromosome 19 variant table ...\n",
      "\t\tMaking chromosome 2 variant table ...\n",
      "\t\tMaking chromosome 20 variant table ...\n",
      "\t\tMaking chromosome 21 variant table ...\n",
      "\t\tMaking chromosome 22 variant table ...\n",
      "\t\tMaking chromosome 3 variant table ...\n",
      "\t@ 3000001 ...\n",
      "\t\tMaking chromosome 4 variant table ...\n",
      "\t\tMaking chromosome 5 variant table ...\n",
      "\t\tMaking chromosome 6 variant table ...\n",
      "\t@ 4000001 ...\n",
      "\t\tMaking chromosome 7 variant table ...\n",
      "\t\tMaking chromosome 8 variant table ...\n",
      "\t\tMaking chromosome 9 variant table ...\n",
      "\t\tMaking chromosome MT variant table ...\n",
      "\t\tMaking chromosome X variant table ...\n",
      "\t@ 5000001 ...\n",
      "\t\tMaking chromosome Y variant table ...\n",
      "Flushing tables and making column indices ...\n",
      "\tchromosome 1 variant table ...\n",
      "\tchromosome 10 variant table ...\n",
      "\tchromosome 11 variant table ...\n",
      "\tchromosome 12 variant table ...\n",
      "\tchromosome 13 variant table ...\n",
      "\tchromosome 14 variant table ...\n",
      "\tchromosome 15 variant table ...\n",
      "\tchromosome 16 variant table ...\n",
      "\tchromosome 17 variant table ...\n",
      "\tchromosome 18 variant table ...\n",
      "\tchromosome 19 variant table ...\n",
      "\tchromosome 2 variant table ...\n",
      "\tchromosome 20 variant table ...\n",
      "\tchromosome 21 variant table ...\n",
      "\tchromosome 22 variant table ...\n",
      "\tchromosome 3 variant table ...\n",
      "\tchromosome 4 variant table ...\n",
      "\tchromosome 5 variant table ...\n",
      "\tchromosome 6 variant table ...\n",
      "\tchromosome 7 variant table ...\n",
      "\tchromosome 8 variant table ...\n",
      "\tchromosome 9 variant table ...\n",
      "\tchromosome MT variant table ...\n",
      "\tchromosome X variant table ...\n",
      "\tchromosome Y variant table ...\n",
      "/home/k/Jumis/guardiome/genome-data/customers/738/variant/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz.hdf5 (File) ''\n",
      "Last modif.: 'Fri May 26 17:03:53 2017'\n",
      "Object Tree: \n",
      "/ (RootGroup) ''\n",
      "/chromosome_10_variants (Table(249020,), shuffle, blosc(1)) ''\n",
      "/chromosome_11_variants (Table(240544,), shuffle, blosc(1)) ''\n",
      "/chromosome_12_variants (Table(231845,), shuffle, blosc(1)) ''\n",
      "/chromosome_13_variants (Table(179803,), shuffle, blosc(1)) ''\n",
      "/chromosome_14_variants (Table(152565,), shuffle, blosc(1)) ''\n",
      "/chromosome_15_variants (Table(146512,), shuffle, blosc(1)) ''\n",
      "/chromosome_16_variants (Table(159467,), shuffle, blosc(1)) ''\n",
      "/chromosome_17_variants (Table(150092,), shuffle, blosc(1)) ''\n",
      "/chromosome_18_variants (Table(138271,), shuffle, blosc(1)) ''\n",
      "/chromosome_19_variants (Table(109788,), shuffle, blosc(1)) ''\n",
      "/chromosome_1_variants (Table(372124,), shuffle, blosc(1)) ''\n",
      "/chromosome_20_variants (Table(144307,), shuffle, blosc(1)) ''\n",
      "/chromosome_21_variants (Table(104367,), shuffle, blosc(1)) ''\n",
      "/chromosome_22_variants (Table(107316,), shuffle, blosc(1)) ''\n",
      "/chromosome_2_variants (Table(401037,), shuffle, blosc(1)) ''\n",
      "/chromosome_3_variants (Table(336097,), shuffle, blosc(1)) ''\n",
      "/chromosome_4_variants (Table(355664,), shuffle, blosc(1)) ''\n",
      "/chromosome_5_variants (Table(285518,), shuffle, blosc(1)) ''\n",
      "/chromosome_6_variants (Table(296731,), shuffle, blosc(1)) ''\n",
      "/chromosome_7_variants (Table(289264,), shuffle, blosc(1)) ''\n",
      "/chromosome_8_variants (Table(243301,), shuffle, blosc(1)) ''\n",
      "/chromosome_9_variants (Table(236641,), shuffle, blosc(1)) ''\n",
      "/chromosome_MT_variants (Table(30,), shuffle, blosc(1)) ''\n",
      "/chromosome_X_variants (Table(106614,), shuffle, blosc(1)) ''\n",
      "/chromosome_Y_variants (Table(23120,), shuffle, blosc(1)) ''\n",
      "\n",
      "Writing ID-to-chromosome dict ...\n",
      "Writing gene-to-chromosome dict ...\n",
      "Reading variant HDF5 ...\n"
     ]
    }
   ],
   "source": [
    "VARIANT_FILE_PATH = '/home/k/Jumis/guardiome/genome-data/customers/738/variant/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz'\n",
    "variant_hdf5 = VariantHDF5(VARIANT_FILE_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading chromosome_11_variants where: ID == b'rs662799' ...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'ALT': 'A',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs662799',\n",
       "  'QUAL': 1181.77,\n",
       "  'REF': 'G',\n",
       "  'effect': 'upstream_gene_variant',\n",
       "  'end': 116792991,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116792991,\n",
       "  'variant_type': 'SNP'}]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_hdf5.get_variants_by_id('rs662799')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading chromosome_11_variants where: gene_name == b'ZPR1' ...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'ALT': 'G',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs2041967',\n",
       "  'QUAL': 1211.77,\n",
       "  'REF': 'A',\n",
       "  'effect': '3_prime_UTR_variant',\n",
       "  'end': 116774433,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116774433,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'A',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs45443305',\n",
       "  'QUAL': 1316.77,\n",
       "  'REF': 'C',\n",
       "  'effect': '3_prime_UTR_variant',\n",
       "  'end': 116774966,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116774966,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'C',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': '.',\n",
       "  'QUAL': 1643.73,\n",
       "  'REF': 'CAAAAAAA',\n",
       "  'effect': '3_prime_UTR_variant',\n",
       "  'end': 116775463,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116775457,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'C',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': '.',\n",
       "  'QUAL': 417.73001,\n",
       "  'REF': 'CAAAAAAAAAAAAAA',\n",
       "  'effect': '3_prime_UTR_variant',\n",
       "  'end': 116775641,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116775628,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'T',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs2160669',\n",
       "  'QUAL': 1472.77,\n",
       "  'REF': 'C',\n",
       "  'effect': '3_prime_UTR_variant',\n",
       "  'end': 116776891,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116776891,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'C',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': '.',\n",
       "  'QUAL': 748.72998,\n",
       "  'REF': 'CAA',\n",
       "  'effect': '3_prime_UTR_variant',\n",
       "  'end': 116777665,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116777664,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'C',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs964184',\n",
       "  'QUAL': 1119.77,\n",
       "  'REF': 'G',\n",
       "  'effect': '3_prime_UTR_variant',\n",
       "  'end': 116778201,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116778201,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'G',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs139753514',\n",
       "  'QUAL': 611.72998,\n",
       "  'REF': 'GA',\n",
       "  'effect': 'downstream_gene_variant',\n",
       "  'end': 116781214,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116781214,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'A',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs6589566',\n",
       "  'QUAL': 1376.77,\n",
       "  'REF': 'G',\n",
       "  'effect': 'downstream_gene_variant',\n",
       "  'end': 116781707,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116781707,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'G',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs7483863',\n",
       "  'QUAL': 1269.77,\n",
       "  'REF': 'A',\n",
       "  'effect': 'downstream_gene_variant',\n",
       "  'end': 116781775,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116781775,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'T',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs2075290',\n",
       "  'QUAL': 900.77002,\n",
       "  'REF': 'C',\n",
       "  'effect': 'downstream_gene_variant',\n",
       "  'end': 116782580,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116782580,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'T',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': '.',\n",
       "  'QUAL': 521.72998,\n",
       "  'REF': 'TA',\n",
       "  'effect': 'downstream_gene_variant',\n",
       "  'end': 116783861,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116783861,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'A',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs74662600',\n",
       "  'QUAL': 842.77002,\n",
       "  'REF': 'G',\n",
       "  'effect': 'downstream_gene_variant',\n",
       "  'end': 116784243,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116784243,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'A',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs10750096',\n",
       "  'QUAL': 1075.77,\n",
       "  'REF': 'C',\n",
       "  'effect': 'upstream_gene_variant',\n",
       "  'end': 116786072,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116786072,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'T',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs2075294',\n",
       "  'QUAL': 836.77002,\n",
       "  'REF': 'G',\n",
       "  'effect': 'upstream_gene_variant',\n",
       "  'end': 116787406,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116787406,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'A',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs662799',\n",
       "  'QUAL': 1181.77,\n",
       "  'REF': 'G',\n",
       "  'effect': 'upstream_gene_variant',\n",
       "  'end': 116792991,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116792991,\n",
       "  'variant_type': 'SNP'}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_hdf5.get_variants_by_gene('ZPR1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading chromosome_11_variants where: (116792991 <= start) & (start <= 116792991) ...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'ALT': 'A',\n",
       "  'CHROM': '11',\n",
       "  'GT': 'None',\n",
       "  'ID': 'rs662799',\n",
       "  'QUAL': 1181.77,\n",
       "  'REF': 'G',\n",
       "  'effect': 'upstream_gene_variant',\n",
       "  'end': 116792991,\n",
       "  'gene_name': 'ZPR1',\n",
       "  'impact': 'MODIFIER',\n",
       "  'pathogenicity': '',\n",
       "  'start': 116792991,\n",
       "  'variant_type': 'SNP'}]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_hdf5.get_variants_by_region(11, 116792991, 116792991)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
