{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '../')\n",
    "from variant.varianthdf5 import VariantHDF5\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading variant HDF5 ...\n",
      "Reading ID-to-chromosome dict ...\n",
      "Reading gene-to-chromosome dict ...\n"
     ]
    }
   ],
   "source": [
    "VARIANT_FILE_PATH = '/home/k/Jumis/guardiome/data/customers/738/variants/738.concat_snp_indel.extract_chr.recode.ann_snpeff.ann_clinvar.rename_chr.vcf.gz'\n",
    "# variant_hdf5 = VariantHDF5(VARIANT_FILE_PATH, reset=True)\n",
    "variant_hdf5 = VariantHDF5(VARIANT_FILE_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading chromosome_11_variants where: ID == b'rs662799' ...\n",
      "No variant classification for: effect=upstream_gene_va & variant_type=SNP & inframe=True.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'ALT': 'A',\n",
       "  'ANN': {0: {'effect': 'upstream_gene_va',\n",
       "    'gene_name': 'ZPR1',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '11',\n",
       "  'ID': 'rs662799',\n",
       "  'POS': 116792991,\n",
       "  'QUAL': 1181.77,\n",
       "  'REF': 'G',\n",
       "  'end': 116792991,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['A', 'A']}},\n",
       "  'start': 116792991,\n",
       "  'variant_type': 'SNP'}]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_hdf5.get_variants_by_id('rs662799')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading chromosome_17_variants where: gene_name == b'TP53' ...\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=DEL & inframe=False.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=DEL & inframe=True.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=INS & inframe=True.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=INS & inframe=True.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=SNP & inframe=True.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=DEL & inframe=False.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=SNP & inframe=True.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=SNP & inframe=True.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=SNP & inframe=True.\n",
      "No variant classification for: effect=downstream_gene_ & variant_type=SNP & inframe=True.\n",
      "No variant classification for: effect=3_prime_UTR_vari & variant_type=DEL & inframe=False.\n",
      "No variant classification for: effect=upstream_gene_va & variant_type=DEL & inframe=False.\n",
      "No variant classification for: effect=upstream_gene_va & variant_type=INS & inframe=False.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'ALT': 'T',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs33975564',\n",
       "  'POS': 7663144,\n",
       "  'QUAL': 1661.73,\n",
       "  'REF': 'TAA',\n",
       "  'end': 7663146,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['T', 'T']}},\n",
       "  'start': 7663145,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'T',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs67905980',\n",
       "  'POS': 7663147,\n",
       "  'QUAL': 1661.73,\n",
       "  'REF': 'TATA',\n",
       "  'end': 7663150,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['T', 'T']}},\n",
       "  'start': 7663148,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'ATCTATCTATCTATCTATCTATCTG',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': '.',\n",
       "  'POS': 7664361,\n",
       "  'QUAL': 469.73001,\n",
       "  'REF': 'A',\n",
       "  'end': 7664362,\n",
       "  'sample': {0: {'GT': '0/1', 'genotype': ['A', 'ATCTATCTATCTATCTATCTATCTG']}},\n",
       "  'start': 7664361,\n",
       "  'variant_type': 'INS'},\n",
       " {'ALT': 'ATCTATCTATCTG',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': '.',\n",
       "  'POS': 7664373,\n",
       "  'QUAL': 334.73001,\n",
       "  'REF': 'A',\n",
       "  'end': 7664374,\n",
       "  'sample': {0: {'GT': '0/1', 'genotype': ['A', 'ATCTATCTATCTG']}},\n",
       "  'start': 7664373,\n",
       "  'variant_type': 'INS'},\n",
       " {'ALT': 'T',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs55745760',\n",
       "  'POS': 7665607,\n",
       "  'QUAL': 1180.77,\n",
       "  'REF': 'C',\n",
       "  'end': 7665607,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['T', 'T']}},\n",
       "  'start': 7665607,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'T',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs34569991',\n",
       "  'POS': 7667260,\n",
       "  'QUAL': 844.72998,\n",
       "  'REF': 'TA',\n",
       "  'end': 7667261,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['T', 'T']}},\n",
       "  'start': 7667261,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'A',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs12949853',\n",
       "  'POS': 7667560,\n",
       "  'QUAL': 1784.77,\n",
       "  'REF': 'G',\n",
       "  'end': 7667560,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['A', 'A']}},\n",
       "  'start': 7667560,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'T',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs4968204',\n",
       "  'POS': 7667611,\n",
       "  'QUAL': 2071.77,\n",
       "  'REF': 'C',\n",
       "  'end': 7667611,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['T', 'T']}},\n",
       "  'start': 7667611,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'G',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs4968186',\n",
       "  'POS': 7667612,\n",
       "  'QUAL': 2071.77,\n",
       "  'REF': 'A',\n",
       "  'end': 7667612,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['G', 'G']}},\n",
       "  'start': 7667612,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'G',\n",
       "  'ANN': {0: {'effect': 'downstream_gene_',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs9894946',\n",
       "  'POS': 7667762,\n",
       "  'QUAL': 1274.77,\n",
       "  'REF': 'A',\n",
       "  'end': 7667762,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['G', 'G']}},\n",
       "  'start': 7667762,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'G',\n",
       "  'ANN': {0: {'effect': '3_prime_UTR_vari',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'CLNDBN': 'Li-Fraumeni_synd',\n",
       "  'CLNSIG': '0',\n",
       "  'ID': 'rs200757381',\n",
       "  'POS': 7668836,\n",
       "  'QUAL': 314.73999,\n",
       "  'REF': 'GA',\n",
       "  'clinvar': 'unknown',\n",
       "  'end': 7668837,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['G', 'G']}},\n",
       "  'start': 7668837,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'C',\n",
       "  'ANN': {0: {'effect': 'sequence_feature',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'LOW',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs1642793',\n",
       "  'POS': 7672246,\n",
       "  'QUAL': 1006.77,\n",
       "  'REF': 'T',\n",
       "  'end': 7672246,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['C', 'C']}},\n",
       "  'start': 7672246,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'CAA',\n",
       "  'ANN': {0: {'effect': 'sequence_feature',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'LOW',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs34308401',\n",
       "  'POS': 7673125,\n",
       "  'QUAL': 151.19,\n",
       "  'REF': 'C',\n",
       "  'end': 7673126,\n",
       "  'sample': {0: {'GT': '0/1', 'genotype': ['C', 'CAA']}},\n",
       "  'start': 7673125,\n",
       "  'variant_type': 'INS'},\n",
       " {'ALT': 'C',\n",
       "  'ANN': {0: {'effect': 'sequence_feature',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'LOW',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': '0.1663,0',\n",
       "  'CHROM': '17',\n",
       "  'CLNDBN': 'not_specified',\n",
       "  'CLNSIG': '2',\n",
       "  'ID': 'rs1625895',\n",
       "  'POS': 7674797,\n",
       "  'QUAL': 1395.77,\n",
       "  'REF': 'T',\n",
       "  'clinvar': 'non-pathogenic',\n",
       "  'end': 7674797,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['C', 'C']}},\n",
       "  'start': 7674797,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'T',\n",
       "  'ANN': {0: {'effect': 'sequence_feature',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'LOW',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs2909430',\n",
       "  'POS': 7675327,\n",
       "  'QUAL': 1016.77,\n",
       "  'REF': 'C',\n",
       "  'end': 7675327,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['T', 'T']}},\n",
       "  'start': 7675327,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'C',\n",
       "  'ANN': {0: {'effect': 'sequence_feature',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'LOW',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs141204613',\n",
       "  'POS': 7675393,\n",
       "  'QUAL': 939.72998,\n",
       "  'REF': 'CTTT',\n",
       "  'end': 7675396,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['C', 'C']}},\n",
       "  'start': 7675394,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'G',\n",
       "  'ANN': {0: {'effect': 'sequence_feature',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'LOW',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': 'rs1794287',\n",
       "  'POS': 7675519,\n",
       "  'QUAL': 1350.77,\n",
       "  'REF': 'A',\n",
       "  'end': 7675519,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['G', 'G']}},\n",
       "  'start': 7675519,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'C',\n",
       "  'ANN': {0: {'effect': 'missense_variant',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODERATE',\n",
       "    'variant_classification': 'Missense_Mutation'}},\n",
       "  'CAF': '0.4571,0',\n",
       "  'CHROM': '17',\n",
       "  'CLNDBN': 'not_specified,CO',\n",
       "  'CLNSIG': '2',\n",
       "  'ID': 'rs1042522',\n",
       "  'POS': 7676154,\n",
       "  'QUAL': 1083.77,\n",
       "  'REF': 'G',\n",
       "  'clinvar': 'non-pathogenic',\n",
       "  'end': 7676154,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['C', 'C']}},\n",
       "  'start': 7676154,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'C',\n",
       "  'ANN': {0: {'effect': 'sequence_feature',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'LOW',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': '0.4135,0',\n",
       "  'CHROM': '17',\n",
       "  'CLNDBN': 'Hereditary_cance',\n",
       "  'CLNSIG': '2',\n",
       "  'ID': 'rs146534833;rs59',\n",
       "  'POS': 7676325,\n",
       "  'QUAL': 1345.73,\n",
       "  'REF': 'CCCCCAGCCCTCCAGGT',\n",
       "  'clinvar': 'non-pathogenic',\n",
       "  'end': 7676341,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['C', 'C']}},\n",
       "  'start': 7676326,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'C',\n",
       "  'ANN': {0: {'effect': 'sequence_feature',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'LOW',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': '0.4151,0',\n",
       "  'CHROM': '17',\n",
       "  'CLNDBN': 'not_specified',\n",
       "  'CLNSIG': '2',\n",
       "  'ID': 'rs1642785',\n",
       "  'POS': 7676483,\n",
       "  'QUAL': 1270.77,\n",
       "  'REF': 'G',\n",
       "  'clinvar': 'non-pathogenic',\n",
       "  'end': 7676483,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['C', 'C']}},\n",
       "  'start': 7676483,\n",
       "  'variant_type': 'SNP'},\n",
       " {'ALT': 'A',\n",
       "  'ANN': {0: {'effect': 'upstream_gene_va',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': '.',\n",
       "  'POS': 7676938,\n",
       "  'QUAL': 401.73001,\n",
       "  'REF': 'AT',\n",
       "  'end': 7676939,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['A', 'A']}},\n",
       "  'start': 7676939,\n",
       "  'variant_type': 'DEL'},\n",
       " {'ALT': 'AAAAC',\n",
       "  'ANN': {0: {'effect': 'upstream_gene_va',\n",
       "    'gene_name': 'TP53',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '17',\n",
       "  'ID': '.',\n",
       "  'POS': 7691209,\n",
       "  'QUAL': 297.73001,\n",
       "  'REF': 'A',\n",
       "  'end': 7691210,\n",
       "  'sample': {0: {'GT': '0/1', 'genotype': ['A', 'AAAAC']}},\n",
       "  'start': 7691209,\n",
       "  'variant_type': 'INS'}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_hdf5.get_variants_by_gene('TP53')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading chromosome_11_variants where: (116792991 <= POS) & (POS <= 116792991) ...\n",
      "No variant classification for: effect=upstream_gene_va & variant_type=SNP & inframe=True.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'ALT': 'A',\n",
       "  'ANN': {0: {'effect': 'upstream_gene_va',\n",
       "    'gene_name': 'ZPR1',\n",
       "    'impact': 'MODIFIER',\n",
       "    'variant_classification': 'Targeted_Region'}},\n",
       "  'CAF': 'None',\n",
       "  'CHROM': '11',\n",
       "  'ID': 'rs662799',\n",
       "  'POS': 116792991,\n",
       "  'QUAL': 1181.77,\n",
       "  'REF': 'G',\n",
       "  'end': 116792991,\n",
       "  'sample': {0: {'GT': '1/1', 'genotype': ['A', 'A']}},\n",
       "  'start': 116792991,\n",
       "  'variant_type': 'SNP'}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_hdf5.get_variants_by_region(11, 116792991, 116792991)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
